---
title: "LC analyses: additional marker genes"
author: "Lukas Weber"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    number_sections: true
    toc_depth: 3
    toc_float:
      collapsed: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Introduction

Analyses for LC project: additional marker genes.


## Data

In this file, we are using the combined NextSeq and MiSeq reads for the LC samples (LC_1 and LC_2).


# Analysis

## Load data

See previous file `2_features_per_spot_LC.Rmd` for more details.

```{r, message=FALSE}
library(here)
library(tidyverse)
library(ggplot2)
library(SingleCellExperiment)
library(Matrix)
library(rjson)
```

```{r}
sample_names <- c("LC_1", "LC_2")
sce_list <- vector("list", length = length(sample_names))
names(sce_list) <- sample_names

for (i in seq_along(sample_names)) {
  
  # ---------
  # load data
  # ---------
  
  # select sample
  sample_name <- sample_names[i]
  
  # path to Space Ranger output files
  if (Sys.info()["sysname"] == "Linux") {
    # files on JHPCE cluster
    dir_outputs <- "/dcl02/lieber/ajaffe/SpatialTranscriptomics/LIBD/locus-c/outputs/NextSeqMiSeq"
  } else if (Sys.info()["sysname"] == "Darwin") {
    # copy of files on Mac laptop
    dir_outputs <- "~/data/locus-c/outputs/NextSeqMiSeq"
  }
  
  # note: using "filtered" barcodes list containing only spots over tissue
  dir_matrix <- file.path(dir_outputs, sample_name, "outs", "filtered_feature_bc_matrix")
  
  # barcodes
  file_barcodes <- file.path(dir_matrix, "barcodes.tsv.gz")
  df_barcodes <- read.csv(file_barcodes, sep = "\t", header = FALSE, 
                          col.names = c("barcode_id"))
  # features
  file_features <- file.path(dir_matrix, "features.tsv.gz")
  df_features <- read.csv(file_features, sep = "\t", header = FALSE, 
                          col.names = c("gene_id", "gene_name", "feature_type"))
  # counts
  file_counts <- file.path(dir_matrix, "matrix.mtx.gz")
  counts <- readMM(file = file_counts)
  
  # spatial coordinates
  dir_spatial <- file.path(dir_outputs, sample_name, "outs", "spatial")
  file_tisspos <- file.path(dir_spatial, "tissue_positions_list.csv")
  df_tisspos <- read.csv(file_tisspos, header = FALSE, 
                         col.names=c("barcode_id", "in_tissue", "array_row", "array_col", 
                                     "pxl_col_in_fullres", "pxl_row_in_fullres"))
  
  # check dimensions
  dim(df_barcodes)
  dim(df_features)
  dim(counts)
  # note df_tisspos contains all spots (not filtered) - need to match later
  dim(df_tisspos)
  
  # image paths
  imageFilePath <- file.path(dir_spatial, c("tissue_hires_image.png", "tissue_lores_image.png"))
  
  # spatial scale factors
  file_scale <- file.path(dir_spatial, "scalefactors_json.json")
  scalefactors <- fromJSON(file = file_scale)
  
  
  # ---------------------------
  # create SingleCellExperiment
  # ---------------------------
  
  # note: check and/or re-order rows to make sure barcode IDs match in df_barcodes and df_tisspos
  dim(df_barcodes)
  dim(df_tisspos)
  ord <- match(df_barcodes$barcode_id, df_tisspos$barcode_id)
  df_tisspos_ord <- df_tisspos[ord, ]
  dim(df_tisspos_ord)
  stopifnot(nrow(df_barcodes) == nrow(df_tisspos_ord))
  stopifnot(all(df_barcodes$barcode_id == df_tisspos_ord$barcode_id))
  
  head(df_barcodes)
  head(df_tisspos_ord)
  
  col_data <- cbind(df_barcodes, df_tisspos_ord[, -1])
  head(col_data)
  
  sce <- SingleCellExperiment(
    rowData = df_features, 
    colData = col_data, 
    assays = c(counts = counts), 
    metadata = list(scalefactors = scalefactors)
  )
  
  sce
  
  # store object
  sce_list[[i]] <- sce
}

sce_list
```


## Calculate scater QC metrics

```{r, message=FALSE}
library(scater)
```

```{r, message=FALSE}
for (i in seq_along(sample_names)) {
  
  # select sample
  sce <- sce_list[[i]]
  
  # identify mitochondrial genes
  is_mito <- grepl("(^MT-)|(^mt-)", rowData(sce)$gene_name)
  table(is_mito)
  rowData(sce)$gene_name[is_mito]
  
  # calculate QC metrics using scater package
  sce <- addPerCellQC(sce, subsets = list(mito = is_mito))
  
  colData(sce)
  
  # store
  sce_list[[i]] <- sce
}
```


## Additional marker genes

Additional marker genes: GAD1, GAD2, PVALB, SST, CORT, 5HT3AR, KIT, VIP, NPY, CRHBP, CALB1, RELN

Also: BDNF, NTRK2

Note: can't find 5HT3AR in gene list

```{r, message=FALSE, fig.width=3.5, fig.height=2.5}
# expression of known marker genes using raw UMI counts

dir_plot <- here("plots", "additional_markers")
if (!dir.exists(dir_plot)) dir.create(dir_plot, recursive = TRUE)

marker_gene_names <- c("GAD1", "GAD2", "PVALB", "SST", "CORT", "KIT", "VIP", "NPY", "CRHBP", "CALB1", "RELN", "BDNF", "NTRK2")

colors <- c("navy", "deeppink")

for (i in seq_along(sample_names)) {
  
  # select sample
  sce <- sce_list[[i]]
  
  for (j in seq_along(marker_gene_names)) {
    
    # identify marker gene
    ix_marker <- which(rowData(sce)$gene_name == marker_gene_names[j])
    colData(sce)$counts_marker <- counts(sce)[ix_marker, ]
    
    
    # plot UMI counts for marker gene
    
    p <- ggplot(as.data.frame(colData(sce)), 
                aes(x = pxl_row_in_fullres, y = pxl_col_in_fullres, color = counts_marker)) + 
      geom_point(size = 0.25) + 
      coord_fixed() + 
      scale_y_reverse() + 
      scale_color_gradient(low = "gray95", high = colors[i]) + 
      ggtitle(paste0(sample_names[i], ": UMI counts for gene ", marker_gene_names[j])) + 
      theme_bw() + 
      theme(panel.grid = element_blank(), 
            axis.title = element_blank(), 
            axis.text = element_blank(), 
            axis.ticks = element_blank())
    
    print(p)
    
    fn <- file.path(dir_plot, paste0("xyplot_counts_", marker_gene_names[j], "_", sample_names[i], ".png"))
    ggsave(fn, width = 3.5, height = 2.5)
  }
}
```


# Session info

```{r, message=FALSE}
sessionInfo()
```

